libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/internal/MapboxNavigationViewApi.kt

```diff
@@ -0,0 +1,105 @@
+package com.mapbox.navigation.dropin.internal
+
+import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.geojson.Point
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.dropin.NavigationViewApi
+import com.mapbox.navigation.ui.app.internal.Store
+import com.mapbox.navigation.ui.app.internal.destination.Destination
+import com.mapbox.navigation.ui.app.internal.destination.DestinationAction
+import com.mapbox.navigation.ui.app.internal.extension.dispatch
+import com.mapbox.navigation.ui.app.internal.navigation.NavigationState
+import com.mapbox.navigation.ui.app.internal.navigation.NavigationStateAction
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutePreviewAction
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutePreviewState
+import com.mapbox.navigation.ui.app.internal.routefetch.RoutesAction
+import com.mapbox.navigation.ui.app.internal.tripsession.TripSessionStarterAction
+import com.mapbox.navigation.ui.app.internal.endNavigation as endNavigationAction
+
+@ExperimentalPreviewMapboxNavigationAPI
+internal class MapboxNavigationViewApi(
+    private val store: Store
+) : NavigationViewApi() {
+
+    override fun fetchRoutes(options: RouteOptions) {
+        store.dispatch(RoutePreviewAction.FetchOptions(options))
+    }
+
+    override fun fetchRoutes(points: List<Point>) {
+        store.dispatch(RoutePreviewAction.FetchPoints(points))
+    }
+
+    override fun setPreviewRoutes(routes: List<NavigationRoute>) {
+        store.dispatch(RoutePreviewAction.Ready(routes))
+    }
+
+    override fun setRoutes(routes: List<NavigationRoute>) {
+        store.dispatch(RoutesAction.SetRoutes(routes))
+    }
+
+    override fun setDestination(point: Point) {
+        store.dispatch(DestinationAction.SetDestination(Destination(point)))
+    }
+
+    override fun enableTripSession() {
+        store.dispatch(TripSessionStarterAction.EnableTripSession)
+    }
+
+    override fun enableReplaySession() {
+        store.dispatch(TripSessionStarterAction.EnableReplayTripSession)
+    }
+
+    override fun isReplayEnabled(): Boolean {
+        return store.state.value.tripSession.isReplayEnabled
+    }
+
+    override fun startFreeDrive() {
+        store.dispatch(RoutesAction.SetRoutes(emptyList()))
+        store.dispatch(RoutePreviewAction.Ready(emptyList()))
+        store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
+    }
+
+    @Throws(IllegalStateException::class)
+    override fun startRoutePreview() {
+        checkDestination()
```

Comment:
We should not check for the existence of destination here. Doing this we are forcing end users to go through `DestinationPreview` state even if they don't want to.

Reply:
Ok. 
QQ: How do you get destination cooridnates from the `NavigationRoute`? Is this correct?

```kotlin
val destination: Point? = routes.first().routeOptions.coordinatesList().lastOrNull()
```

Reply:
That or we could also do `routes.first().directionsResponse.waypoints()?.lastOrNull()`.
@LukasPaczos Can you confirm is there a preference of one over the other?

Reply:
There are 3 ways to represent the destination:
- `routeOptions.coordinatesList().lastOrNull()` which is the requested destination
- `directionsResponse.waypoints()?.lastOrNull()` which is the requested destination snapped to a road network (so it's going to snapped to the road accessible closest to wherever the POI may be)
- `routeOptions.waypointTargetsList().lastOrNull()` which can be an additional representation of a destination that's presumably inaccessible from a road network or shouldn't be guide to directly (like a drop-off point).

Depending on on a scenario, either of these can be used. Additionally, there could be `directionsResponse.waypoints()` injected by the backend for multi-leg routes (charging stations for electric vehicles).

To unblock the PR, I think we should start with using `routes.first().directionsResponse.waypoints()?.lastOrNull()`, as that's guaranteed to be correct, however, it won't allow us to put a destination pin on the POI, the pin will be placed on the road. After that we should cut a ticket to determine what would we actually want to mark with the destination pin and build a short algorithm and data structure that can support it. We'll need a `Destination` type that holds both the routable point (snapped to the road network) and a destination point (which can be used to visualize the target). In this type of a data structure I think that the destination point should be determined like this: `routeOptions.waypointTargetsList().last` (if exists) -> `routeOptions.coordinatesList().last` (if exists) -> `directionsResponse.waypoints().last`.

Reply:
In 1TAP we use `routeOptions.coordinatesList().last()`. I see that Drop-In uses `directionsResponse.waypoints()?.lastOrNull()?.location()` and it doesn't look good in my opinion. As you can see in the video, I tapped on the POI, but the pin was placed in a different location. We already have grey circles denoting the waypoints, so I think we can use destination markers for something else. 

https://user-images.githubusercontent.com/2395284/187981836-240fc55a-2d80-4e2e-9e12-96402fc0624c.mp4

Reply:
Maybe `startRoutePreview` should have a parameter for specifying destination.

Reply:
@Zayankovsky, Could you open a new issue with your proposal?

Reply:
Added a comment here: https://github.com/mapbox/navigation-sdks/issues/1952#issuecomment-1234643487.